roachtest: randomly run with runtime assertions by default

DarrylWong opened the pull request:
This changes the semantics of `t.Cockroach()` to use a binary with runtime assertions enabled by default. Performance tests (indicated by the benchmark TestSpec) will be able to continue using the standard binary, without runtime assertions or metamorphic constants, as usual.

This commit also opts-out other tests that cannot easily be run with runtime assertions or metamorphic constants enabled, most often due to timeouts or metamorphic constant incompatibility.

Resolves #86678.
Informs #94986.

Epic: none

Release note: None

Commit:
roachtest: randomly run with runtime assertions by default

This changes the semantics of `t.Cockroach()` to use a binary with
runtime assertions and metamorphic constants enabled by default.
Performance tests (indicated by the benchmark TestSpec) will continue
using the standard binary, without runtime assertions, as usual.

This commit also opts-out other tests that cannot easily be run with
runtime assertions or metamorphic constants enabled, most often due to
timeouts and metamorphic constant incompatibility.

Resolves #86678.
Informs #94986.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111949)
<!-- Reviewable:end -->

DarrylWong commented on pkg/cmd/roachtest/tests/kvbench.go:
```diff
@@ -96,6 +96,7 @@ func registerKVBenchSpec(r registry.Registry, b kvBenchSpec) {
 		Suites:           registry.ManualOnly,
 		Tags:             registry.Tags("manual"),
 		Owner:            registry.OwnerKV,
+		Benchmark:        true, // TODO: Comments imply this is a benchmark test but this was not set
```
The test name and comments imply this is a benchmark test so I assume the missing benchmark spec is a mistake but if there is a reason why it shouldn't be set let me know.

DarrylWong commented:
Should be RFAL

[Test Run](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12372626?hideProblemsFromDependencies=false&hideTestsFromDependencies=false)

rickystewart commented on pkg/cmd/roachtest/test_impl.go:
```diff
@@ -59,8 +62,12 @@ type failure struct {
-	cockroach          string // path to main cockroach binary
-	cockroachShort     string // path to cockroach-short binary compiled with --crdb_test build tag
+	cockroach      string // path to main cockroach binary
+	cockroachShort string // path to cockroach-short binary compiled with --crdb_test build tag
```
Nit: `cockroachAssertionsEnabled` (or similar, it seems elsewhere we use `cockroachEA` for "enabled assertions") is a better name than `cockroachShort`. Yes, the comment explains the real difference between these two things, but everyone sees the variable name, not everyone sees the comment.

rickystewart commented on build/teamcity/cockroach/nightlies/roachtest_compile_component.sh:
```diff
@@ -83,7 +83,7 @@ case "$component" in
   roachtest)
     # Roachtest binary.
-    bazel_args=(//pkg/cmd/roachtest)
+    bazel_args=(//pkg/cmd/roachtest --crdb_test)
```
What benefit does this give exactly?

DarrylWong replied:
Stealing from what [Stan said:](https://github.com/cockroachdb/cockroach/pull/113113)

> Note, to fully support interop between a roachtest runner, (executing the roachtest binary), and a node, executing the cockroach binary compiled with assertions (i.e., crdb_test), the roachtest binary must also be compiled with assertions to establish the same serialization on the wire. E.g., KVNemesisSeq is used on the wire only under crdb_test builds. The corresponding (sequence) Container type is conditional on the binary being compiled under crdb_test. (See pkg/kv/kvpb/api.proto) Thus, a request serialized from the roachtest runner is compatible with the cockroach binary iff both binaries are built under crdb_test.

The solution here was to just compile all roachtest binaries with crdb_test since it touches very little code, rather than try and figure out a way to have team city selectively use binaries based on what the cockroach binary is compiled with.

DarrylWong commented:
More test failures that expected but looks like all are either failing on master or memory budget issues. `tpch_concurrency/no_streamer` is the only one I'm a little worried about. I keep raising the budget but it keeps failing due to OOM. Looks like its failed in the past in master due to OOM, so I might have to just disable assertions for it if it won't play nice wrt memory.

herkolategan reviewed (commented):
Nice work! Made a few small comments / nits.

On a side note: This may be personal preference, but I find the `StandardCockroach()` vs. `Cockroach()` a bit elusive, until a visit to the doc for each method that explains the concept. I would have maybe opted to rather specify in the test spec that the test is incompatible with assertions, and remove the ability to select `StandardCockroach()`. The test logic can then select the binary based on the test spec (e.g., opt out of EA) and this would also make a centralized place (test spec) to add the reason for the test being incompatible.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-Ni4vK6uDg_W-EhtmtdR:b-gkyfw5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, @rickystewart, and @srosenberg)

___
*[`build/teamcity/cockroach/nightlies/roachtest_nightly_aws.sh` line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni4rQF50evXMq1VwhUY:-Ni4rQF50evXMq1VwhUZ:b-u37i4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/540a18fba27db4297178a758a5b763d32eb4601d/build/teamcity/cockroach/nightlies/roachtest_nightly_aws.sh#L10)):*
> ```Shell
> source "$dir/teamcity-bazel-support.sh"  # For run_bazel
> 
> BAZEL_SUPPORT_EXTRA_DOCKER_ARGS="-e LITERAL_ARTIFACTS_DIR=$root/artifacts -e AWS_ACCESS_KEY_ID -e AWS_ACCESS_KEY_ID_ASSUME_ROLE -e AWS_KMS_KEY_ARN_A -e AWS_KMS_KEY_ARN_B -e AWS_KMS_REGION_A -e AWS_KMS_REGION_B -e AWS_ROLE_ARN -e AWS_SECRET_ACCESS_KEY -e AWS_SECRET_ACCESS_KEY_ASSUME_ROLE -e BUILD_VCS_NUMBER -e CLOUD -e COCKROACH_DEV_LICENSE -e TESTS -e COUNT -e GITHUB_API_TOKEN -e GITHUB_ORG -e GITHUB_REPO -e GOOGLE_EPHEMERAL_CREDENTIALS -e SLACK_TOKEN -e TC_BUILDTYPE_ID -e TC_BUILD_BRANCH -e TC_BUILD_ID -e TC_SERVER_URL -e SELECT_PROBABILITY -e COCKROACH_RANDOM_SEED -e ROACHTEST_ASSERTIONS_ENABLED_SEED" \
> ```

We also have `roachtest_nightly_azure.sh` to which these args can be added for future use.
___
*[`pkg/cmd/roachtest/tests/multitenant_utils.go` line 275 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni4yhX_3NU8hJq483Ff:-Ni4yhX_3NU8hJq483Fg:bl4cc7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/540a18fba27db4297178a758a5b763d32eb4601d/pkg/cmd/roachtest/tests/multitenant_utils.go#L275)):*
> ```Go
> 	errCh := make(chan error, 1)
> 	go func() {
> 		// Set the same random seed for every tenant; this picked up by
> ```

nit: `this is picked up`...


<!-- Sent from Reviewable.io -->

DarrylWong reviewed (commented):
>specify in the test spec

I have #113301 drafted up which moves the cockroach binary `Put()` calls into the roachtest setup phase and removes the need for (almost) every test to manually call `Put()`. Along with that is a `CockroachBinary` option which can be set to `StandardCockroach` which would upload `StandardCockroach()` instead. Names could be tweaked to be say `DisableAssertions` instead to make it more clear and it can be a TestSpec instead of ClusterSpec option but I think the general idea is what you're asking for.

Decided to have this as a followup after talking to Renato and Stan since enough tests are being touched in this PR already and #113301 changes every roachtest. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-Ni55BkgBXGf-qm1UnXF:b-km7gev)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @rickystewart, and @srosenberg)

___
*[`build/teamcity/cockroach/nightlies/roachtest_nightly_aws.sh` line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni4rQF50evXMq1VwhUY:-Ni59wEV2kIYhUCTylyB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/540a18fba27db4297178a758a5b763d32eb4601d/build/teamcity/cockroach/nightlies/roachtest_nightly_aws.sh#L10)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

We also have `roachtest_nightly_azure.sh` to which these args can be added for future use.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
Excellent! Thanks for taking this to the finish line! :checkered_flag: 

A couple of questions:

* I already mentioned this in the comments above, but is there a specific reason why we're skipping runtime checks in a few tests that legitimately break? IMO, we should let them break naturally in the nightlies as they point to real issues. Before we merge, we can send a heads-up to relevant teams on Slack so that they are aware the test may break soon and point them to the detailed issues you wrote.
* I'm confused about `schemachange/bulkingest`: the code in this PR suggests the timeout was bumped to 60mins, but I'm looking at the latest CI build on this branch and I see that it timed out after 2h. Are we confident this test passes?
* Would be interesting to run a full nightly build in this branch forcing metamorphic builds for _all_ tests, to reveal some flakes we might not have noticed. Did you run a build like that? I'm interested to see the outcome.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-Ni5p7m_6KUYuHTS6SU7:bmjxq7c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @rickystewart, and @srosenberg)

___
*[`pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go` line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5mM3ICxTswyaJCwKE:-Ni5mM3ICxTswyaJCwKF:b-axbuww) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go#L140)):*
> ```Go
> 	// Run with standard binary as older versions retrieved through roachprod stage
> 	// are not currently available with crdb_test enabled.
> 	// TODO(DarrylWong): Compile older versions with crdb_test flag.
> ```

A thought: I wonder if running the last upgrade (from previous to current version) with assertions enabled would work. We would need to disable metamorphic constants (`COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING`), but at least we would still exercise runtime assertions.

Not for this PR, mostly a note-to-self as something to explore in the near future.
___
*[`pkg/cmd/roachtest/tests/asyncpg.go` line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5o_ZrCo442KSFc1zs:-Ni5o_ZrCo442KSFc1zt:bttt10h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/asyncpg.go#L54)):*
> ```Go
> 		// TODO(DarrylWong): Use a metamorphic constants exclusion list instead.
> 		// See: https://github.com/cockroachdb/cockroach/issues/113164
> 		c.Put(ctx, t.StandardCockroach(), "./cockroach", c.All())
> ```

In `pgjdbc`, we're forcing the value of this session var -- can't we do this here too?
___
*[`pkg/cmd/roachtest/tests/multitenant_distsql.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5nZK35T8OUHnSWG75:-Ni5nZK35T8OUHnSWG76:bwoomki) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/multitenant_distsql.go#L65)):*
> ```Go
> 	// See: https://github.com/cockroachdb/cockroach/issues/113170
> 	// TODO(DarrylWong): Enable runtime assertions once issue is addressed.
> 	c.Put(ctx, t.StandardCockroach(), "./cockroach")
> ```

We shouldn't skip runtime assertions because this test fails IMO: it's precisely why we're doing this! I think the issue will get a lot more eyes when it is reported as a roachtest failure. My reading of the issue you created is that it is a legitimate bug.
___
*[`pkg/cmd/roachtest/tests/multitenant_utils.go` line 281 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5oJTGEqWhnHqw3XmI:-Ni5oJTGEqWhnHqw3XmJ:byvyd2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/multitenant_utils.go#L281)):*
> ```Go
> 		errCh <- c.RunE(tenantCtx, node,
> 			append(append(append([]string{}, envVars...), binary, "mt", "start-sql"), args...)...,
> 		//append(append(env, binary, "mt", "start-sql"), args...)...,
> ```

Should we delete this?
___
*[`pkg/cmd/roachtest/tests/split.go` line 434 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5os1QDQThRweUTO70:-Ni5os1QDQThRweUTO71:b94ga7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/split.go#L434)):*
> ```Go
> 	// See: https://github.com/cockroachdb/cockroach/issues/112664
> 	// TODO(DarrylWong): once the above issue is resolved, use t.Cockroach()
> 	c.Put(ctx, t.StandardCockroach(), "./cockroach", crdbNodes)
> ```

Same comment re: letting this break.


<!-- Sent from Reviewable.io -->

DarrylWong reviewed (commented):
>but is there a specific reason why we're skipping runtime checks in a few tests that legitimately break?

Specific comments on `kv/split` tests below but in general no there isn't and I agree with the sentiment we should let them fail.

>I'm confused about `schemachange/`bulkingest`

I want to say the connection failure was just an infra flake since I can't repro it with the same seed and it's never failed before w/ runtime assertions. But I also don't see anything that jumps out to me in the logs so I don't want to write it off just yet. Let me dig into this some more.

>PR suggests the timeout was bumped to 60mins

The duration of how long to run the workload was bumped to 60 mins, the timeout is calculated as 2 * duration.

>Would be interesting to run a full nightly build in this branch forcing metamorphic builds for _all_ tests, to reveal some flakes we might not have noticed. Did you run a build like that? I'm interested to see the outcome.

Yeah, that's actually what I was doing most of the time. Here's two runs where I ran all of them [#1](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12119799?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true) and [#2](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12174418?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildTestsSection=true). Unfortunately you can see they timed out, which is why I reverted it back for the final few runs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-Ni5tblS9CNkqexy88sM:b-tv3yzu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @rickystewart, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/asyncpg.go` line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5o_ZrCo442KSFc1zs:-Ni64Eqz4a96M-eD6Pn8:b-uoh38r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/asyncpg.go#L54)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

In `pgjdbc`, we're forcing the value of this session var -- can't we do this here too?
</blockquote></details>

The `asyncpg` tests drops the entire database several times throughout the test which also drops session vars. So even setting the default session value like in `pgjdbc` isn't enough.
___
*[`pkg/cmd/roachtest/tests/multitenant_distsql.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5nZK35T8OUHnSWG75:-Ni64m9C86yGsHs6P1Ps:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/multitenant_distsql.go#L65)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

We shouldn't skip runtime assertions because this test fails IMO: it's precisely why we're doing this! I think the issue will get a lot more eyes when it is reported as a roachtest failure. My reading of the issue you created is that it is a legitimate bug.
</blockquote></details>

SGTM
___
*[`pkg/cmd/roachtest/tests/multitenant_utils.go` line 281 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5oJTGEqWhnHqw3XmI:-Ni6DBsi9YlBt2mBDNCF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/multitenant_utils.go#L281)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Should we delete this?
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/split.go` line 434 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5os1QDQThRweUTO70:-Ni64syD0m-Bj7uQ6yIG:b-p8lxcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/split.go#L434)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Same comment re: letting this break.
</blockquote></details>

The root cause of this issue was found and from my understanding it's not really an urgent issue. Also from what I've been told the faulty QPS function is deprecated anyway. When we have a metamorphic exclusion list, we can just disable `kv-batch-size` and let the other constants/assertions be on.

I'm not opposed to letting this break, I guess I'm more wondering what we do if KV determines the bug isn't worth fixing? Switch it back to `t.StandardCockroach()`?


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> Specific comments on `kv/split` tests below

:+1: sounds good, the exception seems reasonable.

> I want to say the connection failure was just an infra flake since I can't repro it

I think the connection failure is a red herring, the test timed out after 2h.

> Here's two runs where I ran all of them #1 and #2

These are full runs, but are tests running with metamorphic builds all the time? I don't see any custom parameters to force that. Also, when I look at the logs for some tests in that #2 link, I only see the `Running with runtime assertions enabled on global seed` message on some tests, not all.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-Ni6bqkO7T8RiaKOlk8M:buuxv6u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @rickystewart, and @srosenberg)

___
*[`pkg/cmd/roachtest/test_impl.go` line 166 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni6cZ7v3BwRpPQDReOt:-Ni6cZ7v3BwRpPQDReOu:b-xmoebe) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f3709ed8c224e72a14a6223083ea2ccf90ad80/pkg/cmd/roachtest/test_impl.go#L166)):*
> ```Go
> 			// CI build, but we can't assume it exists in every roachtest call.
> 			if path := t.RuntimeAssertionsCockroach(); path != "" {
> 				t.l.Printf("Runtime assertions enabled")
> ```

For instant knowledge of whether we are running with assertions enabled or not, we could add a log message to both branches.
___
*[`pkg/cmd/roachtest/tests/asyncpg.go` line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5o_ZrCo442KSFc1zs:-Ni6cjTA8eMsQoZZY6TM:b-mbsm7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/asyncpg.go#L54)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

The `asyncpg` tests drops the entire database several times throughout the test which also drops session vars. So even setting the default session value like in `pgjdbc` isn't enough.
</blockquote></details>

Got it, thanks. Would it make sense to run these tests with assertions enabled, but disabling metamorphic constants?


<!-- Sent from Reviewable.io -->

DarrylWong reviewed (commented):
>I think the connection failure is a red herring, the test timed out after 2h.

Yeah after running it 5+ times I finally got it to timeout on creating indexes... Seems like an actual issue and not a flake but seems weird why it doesn't consistently fail on the same seed. Will have to investigate more.

>These are full runs, but are tests running with metamorphic builds all the time?

Yeah I did it in an admittedly not very obvious way of just temporarily replacing `t.StandardCockroach()` with the runtime binary in the code itself partly because I was tired of typing in a flag every time. I could do another run if you want, more testing wouldn't hurt but it will probably time out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-Ni6g5DXCMnc5erjHuzr:b-egkyr9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @rickystewart, and @srosenberg)

___
*[`pkg/cmd/roachtest/test_impl.go` line 166 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni6cZ7v3BwRpPQDReOt:-Ni6hOhd5ceiJ2cnOKvs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10f3709ed8c224e72a14a6223083ea2ccf90ad80/pkg/cmd/roachtest/test_impl.go#L166)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

For instant knowledge of whether we are running with assertions enabled or not, we could add a log message to both branches.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/asyncpg.go` line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5o_ZrCo442KSFc1zs:-Ni6hyOv2SFR8vweb_Ub:b-rwsjbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/asyncpg.go#L54)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Got it, thanks. Would it make sense to run these tests with assertions enabled, but disabling metamorphic constants?
</blockquote></details>

I was under the impression that just disabling metamorphic constants wasn't easily doable in roachtests since they get enabled on init. If it's simple functionality I'd be happy to try adding, but otherwise I wonder how much benefit it would be over a metamorphic exclusion list .


<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
> temporarily replacing `t.StandardCockroach()` with the runtime binary in the code itself

Ah, I see it in the commit now!

> Seems like an actual issue

Could be, the last we hear from the index creation is

```
executing declarative schema change PostCommitPhase stage 2 of 7 with 1 BackfillType op (rollback=false) for CREATE INDEX
```

2h before the test timed out, so it got stuck there for a really long time. We should probably let sql-foundations investigate this issue when it occurs on master, as it doesn't seem related to anything we're doing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-NiAF8hh4ltLKwC08QU4:bphijvu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @rickystewart, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/asyncpg.go` line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5o_ZrCo442KSFc1zs:-NiAFooe5Jp59_1in-lJ:b-3iynxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/asyncpg.go#L54)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

I was under the impression that just disabling metamorphic constants wasn't easily doable in roachtests since they get enabled on init. If it's simple functionality I'd be happy to try adding, but otherwise I wonder how much benefit it would be over a metamorphic exclusion list .
</blockquote></details>

You can set `COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=true` when starting the cockroach process, and that will disable use of metamorphic constants in CRDB. Runtime checks (gated with `buildutil.CrdbTestBuild`) will still run.

I'm saying there's still value in running tests with these runtime checks (but without changing any `constants`) instead of always using "standard cockroach" in cases like `asyncpg`. But this can be done in follow-up work too.


<!-- Sent from Reviewable.io -->

DarrylWong reviewed (commented):
>We should probably let sql-foundations investigate this issue when it occurs on master,

Sounds good to me

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-NiAaJpZBnOXSqy21NVG:bb73djm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @rickystewart, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/asyncpg.go` line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-Ni5o_ZrCo442KSFc1zs:-NiAQi2Z16i6L_2vwyTl:b-l2qtke) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c00c4deaef51a93200e2bd872b3f01bf6db6bea/pkg/cmd/roachtest/tests/asyncpg.go#L54)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

You can set `COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=true` when starting the cockroach process, and that will disable use of metamorphic constants in CRDB. Runtime checks (gated with `buildutil.CrdbTestBuild`) will still run.

I'm saying there's still value in running tests with these runtime checks (but without changing any `constants`) instead of always using "standard cockroach" in cases like `asyncpg`. But this can be done in follow-up work too.
</blockquote></details>

Yeah that seems easy enough. In that case, we should only be using `StandardCockroach` for excessive memory usage and timeouts.

Added it in for `asyncpg`, `kv/split` and `copy/bank/rows=1000000` and made them all `t.Cockroach()` again. Initial runs seem good.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-NiFGcWACMKVkjWScB6T:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @rickystewart, and @srosenberg)



<!-- Sent from Reviewable.io -->

renatolabs reviewed (commented):
Something I was thinking about: I think it shouldn't be too difficult to add a "cluster param" in [1] indicating whether the run used metamorphic builds or not. I think it would be nice to give teams an instant way to know if we used metamorphic builds without even having to click through TC artifacts. Thoughts?

[1] https://github.com/cockroachdb/cockroach/blob/cedd409f2115655efe2c529f6052f3056673b277/pkg/cmd/roachtest/github.go#L207

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-NiFHum_C-cPKOSxkVuq:bk63xoe)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @rickystewart, and @srosenberg)



<!-- Sent from Reviewable.io -->

DarrylWong committed:
roachtest: add metamorphicbuild label to github posting

Adds a label indicating that a roachtest failure was
on a metamorphic build on github issue posts.

DarrylWong reviewed (commented):
Done, but unfortunately it's not the cleanest. The binary used is decided at runtime so it's not available through testSpec like the other params are. I had to pass it in as a separate argument which seems a little messy.

It should work but I'm happy to revert the commit and go back to the drawing board if you have any better suggestions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111949#-:-NiFWODsBVMpAuscMdCK:b-5tokf1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @renatolabs, @rickystewart, and @srosenberg)



<!-- Sent from Reviewable.io -->

mikeCRL reviewed (approved):
Confirmed no impacts to docs dependencies within the relevant files.

renatolabs commented:
> It should work but I'm happy to revert the commit and go back to the drawing board

No, what you did is kind of what I had in mind originally. There's no elegant way to aggregate random decision made at test run-time and solving that problem is out of scope for this PR. Given the value that this brings (immediately knowing if we used metamorphic binaries), I think this is a good trade-off.

DarrylWong commented:
TFTRs!

bors r=renatolabs, mikeCRL, rickystewart, herkolategan
